src/silo/common/lineage_tree.cpp

```diff
@@ -0,0 +1,242 @@
+#include "silo/common/lineage_tree.h"
+
+#include <fmt/format.h>
+#include <spdlog/spdlog.h>
+#include <boost/algorithm/string/join.hpp>
+
+#include "silo/common/panic.h"
+#include "silo/preprocessing/preprocessing_exception.h"
+
+namespace silo::common {
+
+namespace {
+
+std::string edgesToString(
+   const std::vector<Idx>& ids,
+   const BidirectionalMap<std::string>& lookup
+) {
+   std::ostringstream oss;
+   for (size_t i = 0; i < ids.size(); ++i) {
+      oss << lookup.getValue(ids[i]);
+      if (i != ids.size() - 1) {
+         oss << " -> ";
+      }
+   }
+   return oss.str();
+}
+
+class Graph {
+   size_t number_of_vertices;
+   std::vector<std::vector<Idx>> adjacency_list;
+
+   std::optional<std::vector<Idx>> findCycleWitnessFromStartVertex(
+      Idx start_vertex,
+      std::vector<bool>& visited
+   ) const;
+
+  public:
+   [[nodiscard]] std::optional<std::vector<Idx>> getCycle() const;
+
+   explicit Graph(size_t number_of_vertices);
+   void addEdge(Idx vertex_from, Idx vertex_to);
+};
+
+Graph::Graph(size_t number_of_vertices)
+    : number_of_vertices(number_of_vertices),
+      adjacency_list(number_of_vertices) {}
+
+void Graph::addEdge(Idx vertex_from, Idx vertex_to) {
+   ASSERT_LT(vertex_from, number_of_vertices);
+   ASSERT_LT(vertex_to, number_of_vertices);
+   adjacency_list.at(vertex_from).emplace_back(vertex_to);
+}
+
+std::optional<std::vector<Idx>> Graph::findCycleWitnessFromStartVertex(
+   Idx start_vertex,
+   std::vector<bool>& visited
+) const {
+   std::vector<Idx> recursion_stack;
+   std::vector<bool> in_recursion_stack(number_of_vertices);
```

Comment:
Do we really need this? Usages can probably be replaced by `recursion_stack.contains()`?

I think we can neglect the time that this algorithm needs compared to the other expensive things that SILO does. IMO performance optimization is irrelevant here, I'd rather simplify the code.

Reply:
The usual "cycle-finding" algorithm actually only has the `in_recursion_stack` and `visited` vectors. They are most commonly referred to as `black` and `grey` vertices. I added the recursion_stack to 1st defend against stack-overflows and more importantly 2nd to show the witness cycle in the error message.

I think it can actually become quite resource consuming. I added a testcase with only 16'000 lineages which go from 32ms to 1.5seconds user-time when using your suggested change

Reply:
Ok, that's more than I expected. Then let's leave it as it is. (Still 1,5 s is quite few compared to the rest of the preprocessing.)